<a href="https://colab.research.google.com/github/sapinspys/lambda-ds-precourse/blob/master/LSDS_Intro_Assignment_8_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School, Intro to Data Science, Day 8 — Classification!

## Assignment

Run this cell to load the Titanic data:

In [0]:
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
train, test = train_test_split(sns.load_dataset('titanic').drop(columns=['alive']), random_state=0)
target = 'survived'

Then, train a [Logistic Regression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict_proba), [Decision Tree](http://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html), or [Random Forest](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) model. Use any features and parameters you want. 

Try to get better than 78.0% accuracy on the test set! (This is not required, but encouraged.)

Do refer to the lecture notebook — but try not to copy-paste.

> You must type each of these exercises in, manually. If you copy and paste, you might as well not even do them. The point of these exercises is to train your hands, your brain, and your mind in how to read, write, and see code. If you copy-paste, you are cheating yourself out of the effectiveness of the lessons. —*[Learn Python the Hard Way](https://learnpythonthehardway.org/book/intro.html)*

After this, you may want to try [Kaggle's Titanic challenge](https://www.kaggle.com/c/titanic)!

In [48]:
train.head()

survived  pclass     sex   age  ...  adult_male  deck  embark_town  alone
105         0       3    male  28.0  ...        True   NaN  Southampton   True
68          1       3  female  17.0  ...       False   NaN  Southampton  False
253         0       3    male  30.0  ...        True   NaN  Southampton  False
320         0       3    male  22.0  ...        True   NaN  Southampton   True
706         1       2  female  45.0  ...       False   NaN  Southampton   True

[5 rows x 14 columns]

In [49]:
corr = train.corr()
corr.style.background_gradient(cmap='coolwarm')

## Logistic Regression: Attempt 1

In [50]:
from sklearn.linear_model import LogisticRegression

features = ["adult_male", "alone"]

model = LogisticRegression()
model.fit(train[features], train[target])

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [51]:
from sklearn.metrics import accuracy_score

y_true = train[target]
y_pred = model.predict(train[features])
print('Train accuracy:', accuracy_score(y_true, y_pred))

y_true = test[target]
y_pred = model.predict(test[features])
print('Test accuracy:', accuracy_score(y_true, y_pred))

Train accuracy: 0.7949101796407185
Test accuracy: 0.7713004484304933


## Decision Tree: Attempt 2

Our first attempt had almost 78% accuracy! Let's see if we can do better using a decision tree model.


In [52]:
from sklearn.tree import DecisionTreeClassifier

features = ["adult_male", "alone"]

model = DecisionTreeClassifier(max_depth=2)
model.fit(train[features], train[target])

y_true = train[target]
y_pred = model.predict(train[features])
print('Train accuracy:', accuracy_score(y_true, y_pred))

y_true = test[target]
y_pred = model.predict(test[features])
print('Test accuracy:', accuracy_score(y_true, y_pred))

Train accuracy: 0.7949101796407185
Test accuracy: 0.7713004484304933


# Multi-Model: Attempt 3

Same results.. let's try different feature configurations

In [53]:
# Encoding sex
train['female'] = train.sex == 'female'
test['female'] = test.sex == 'female'

train[['sex', 'female']].head()

sex  female
105    male   False
68   female    True
253    male   False
320    male   False
706  female    True

In [54]:
# Encoding class
print(f"{train['class'].value_counts()}\n")

class_ranks = {"First": 1, "Second": 2, "Third": 3}
train["class"] = train["class"].map(class_ranks)
test["class"] = test["class"].map(class_ranks)

train.head()

Third     367
First     163
Second    138
Name: class, dtype: int64



survived  pclass     sex   age  ...  deck  embark_town  alone female
105         0       3    male  28.0  ...   NaN  Southampton   True  False
68          1       3  female  17.0  ...   NaN  Southampton  False   True
253         0       3    male  30.0  ...   NaN  Southampton  False  False
320         0       3    male  22.0  ...   NaN  Southampton   True  False
706         1       2  female  45.0  ...   NaN  Southampton   True   True

[5 rows x 15 columns]

In [55]:
# Imputing age

train.age.fillna(train.age.mean(), inplace=True)
test.age.fillna(test.age.mean(), inplace=True)

print(train.age.isnull().sum())
print(test.age.isnull().sum())

0
0


In [56]:
#Logistic Regression 

features = ["class", "female", "age"]

model = LogisticRegression()
model.fit(train[features], train[target])

y_true = train[target]
y_pred = model.predict(train[features])
print('Train accuracy:', accuracy_score(y_true, y_pred))

y_true = test[target]
y_pred = model.predict(test[features])
print('Test accuracy:', accuracy_score(y_true, y_pred))

Train accuracy: 0.7964071856287425
Test accuracy: 0.7757847533632287


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [57]:
# Decision Tree

model = DecisionTreeClassifier(max_depth=2)
model.fit(train[features], train[target])

y_true = train[target]
y_pred = model.predict(train[features])
print('Train accuracy:', accuracy_score(y_true, y_pred))

y_true = test[target]
y_pred = model.predict(test[features])
print('Test accuracy:', accuracy_score(y_true, y_pred))

Train accuracy: 0.7949101796407185
Test accuracy: 0.7757847533632287


# Random Forest: Attempt 4

Same results... will this be the trump card that takes us above 78% test accuracy?

In [59]:
from sklearn.ensemble import RandomForestClassifier

features = ["class", "female", "age"]

model = RandomForestClassifier(n_estimators=10,max_depth=2)
model.fit(train[features], train[target])

y_true = train[target]
y_pred = model.predict(train[features])
print('Train accuracy:', accuracy_score(y_true, y_pred))

y_true = test[target]
y_pred = model.predict(test[features])
print('Test accuracy:', accuracy_score(y_true, y_pred))

Train accuracy: 0.7994011976047904
Test accuracy: 0.820627802690583


### Wooo! Our random forest model returns 82%  test accuracy using class, female, and age features.